### ISSUE TITLE
How to "programmatically" set external proxy per request?

### ISSUE THREAD

**Issue by seidnerj** [ProxymanApp/Proxyman#1844]
In my scenario - I have a process that has multiple sessions, each using a different (but fixed per session) proxy server. This is fine in general, I simply set the code to use a specific proxy server per session. 

I need to see what's happening during the requests done by any given session, for this I use Proxyman. However, now begs the question how do I replicate this same behavior while maintaining the above logic. If I want to use Proxyman, I need to set the proxy server to Proxyman's address, I can also set a specific external proxy, but this isn't enough.

I guess I could use multiple processes of Proxyman, in theory, but I'm not sure if that would work, nor is this a desired experience from a usability standpoint.

Ideally this could provided to Proxyman, per request, by some kind of special header, e.g. "X-Upstream-Proxy" with values such as "http://user:pass@someproxy.com:1312" or "socks5h://user:pass@socksproxy.com:8080".

Is there a way to do something like this currently? if there isn't, is there some kind of workaround that could be implemented to achieve the aforementioned functionality?

Thanks in advance

**Comment by NghiaTranUIT** [comment]
Hey @seidnerj unfortunately, this feature (External Proxy per requests) is not available yet.

Proxyman only supports the [External Proxy](https://docs.proxyman.io/advanced-features/external-proxy) for all requests that go through Proxyman.

I'm not sure how to workaround since macOS allows 1 Proxyman instance at a time. You can't open multiple Proxyman apps.

**Comment by seidnerj** [comment]
I found a workaround for this. I wrote a small mitmproxy add-on that routes a request to a specific upstream proxy based on a custom header such as I suggested above. I then set Proxyman to use this special proxy as the "External Proxy" so all the traffic first passes through Proxyman then proceeds to be routed to the desired upstream proxy server. 

I posted the solution here:

https://stackoverflow.com/questions/77450814/how-to-dynamically-route-and-authenticate-upstream-proxies-in-mitmproxy-based-on
